newsletter-bundle/src/Event/SendNewsletterEvent.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\NewsletterBundle\Event;
+
+use Symfony\Contracts\EventDispatcher\Event;
+
+class SendNewsletterEvent extends Event
+{
+    private string $recipientAddress;
+    private string $text;
+    private string $html;
+    private bool $preventSubmission = false;
+    private bool $allowHtml = true;
+    private array $recipientData = [];
+    private array $newsletterData = [];
+
+    public function __construct(string $recipientAddress, string $text, string $html = '')
+    {
+        $this->recipientAddress = $recipientAddress;
+        $this->text = $text;
+        $this->html = $html;
+    }
+
+    public function getRecipientAddress(): string
+    {
+        return $this->recipientAddress;
+    }
+
+    public function setRecipientAddress(string $recipientAddress): self
+    {
+        $this->recipientAddress = $recipientAddress;
+
+        return $this;
+    }
+
+    public function preventsSubmission(): bool
+    {
+        return $this->preventSubmission;
+    }
+
+    public function setPreventSubmission(bool $preventSubmission): void
+    {
+        $this->preventSubmission = $preventSubmission;
+    }
+
+    public function getText(): string
+    {
+        return $this->text ?? '';
+    }
+
+    public function setText(string $text): self
+    {
+        $this->text = $text;
+
+        return $this;
+    }
+
+    public function getHtml(): string
+    {
+        return $this->html ?? '';
+    }
+
+    public function setHtml(string $html): self
+    {
+        $this->html = $html;
+
+        return $this;
+    }
+
+    public function allowsHtml(): bool
```

Comment:
IMHO we should adopt one of the following naming strategies here:

### 1. `is()` and `set()`

```
public function isAllowHtml(): bool;
public function setAllowHtml(bool $allowHtml);
```

As e.g. done in Composer with the `isClassMapAuthoritative()` and `setClassMapAuthoritative()` methods.

### 2. `is()` and `enable()` + `disable()`

```
public function isHtml(): bool;
public function allowHtml();
public function disallowHtml();
```

As e.g. done in Symfony with the `isDebug()`, `enableDebug()` and `disableDebug()` methods.

### 3. Default setter and getter

```
public function getAllowHtml(): bool;
public function setAllowHtml(bool $allowHtml);
```

As e.g. done in Doctrine entities.

### 4. Speaking methods

```
public function allowsHtml(): bool;
public function allowHtml(bool $allowHtml);
```

However, I have not found an example for this, so I am not sure how common an approach this is.

Reply:
In this specific case I think it's not about allowing something but sending a HTML version along with the text or not? So why not remove this extra property and just allow a nullable parameter in `setHtml(?string $html)`?

Reply:
Thanks for the detailed overview!

I see your point and have adjusted the naming in https://github.com/contao/contao/pull/3489/commits/f1af37d81ae3d7f0d9d755b03f20131f36e29a75 to:
* `isSubmissionPrevented` and `setSubmissionPrevented`
* `isHtmlAllowed` and `setHtmlAllowed`

We could get away with saying `setHtml(null)` is the same as `setHtmlAllowed(false)` for most use cases. However, it wouldn't be possible for the newsletter to be initialized without HTML (`html = null`) but allowing HTML to be added by listeners – and the other way around: for an event to inform all following listeners that HTML is disallowed. The latter case could be handled via proper listener priority (last listener calls `setHtml(null)` but in my opinion, an additional flag is more meaningful here (and prevents possible questions like is `setHtml(null)` the same as `setHtml('')`).